File: Algorithm/Ford-Fulkerson/Edmonds-Karp/atcoder/A68/A68.js
const fs = require('fs');

const input = fs.readFileSync('/dev/stdin', 'utf8').trim().split('\n');
const [N, M] = input[0].split(' ').map(Number);

const graph = Array.from({ length: N + 1 }, () => []); // 隣接リスト
const capacity = Array.from({ length: N + 1 }, () => Array(N + 1).fill(0));

// グラフ構築
for (let i = 1; i <= M; i++) {
    const [a, b, c] = input[i].split(' ').map(Number);
    graph[a].push(b);
    graph[b].push(a); // 逆辺も必要（残余グラフ）
    capacity[a][b] += c;
}

function bfs(level, parent, source, sink) {
    level.fill(-1);
    level[source] = 0;
    const queue = [source];

    while (queue.length > 0) {
        const current = queue.shift();

        for (const next of graph[current]) {
            if (level[next] === -1 && capacity[current][next] > 0) {
                level[next] = level[current] + 1;
                parent[next] = current;
                queue.push(next);
                if (next === sink) return true;
            }
        }
    }
    return false;
}

function maxFlow(source, sink) {
    let flow = 0;
    const parent = Array(N + 1);
    const level = Array(N + 1);

    while (bfs(level, parent, source, sink)) {
        // 最小容量を見つける
        let pathFlow = Infinity;
        let v = sink;
        while (v !== source) {
            const u = parent[v];
            pathFlow = Math.min(pathFlow, capacity[u][v]);
            v = u;
        }

        // 容量を減らす・増やす（残余グラフ更新）
        v = sink;
        while (v !== source) {
            const u = parent[v];
            capacity[u][v] -= pathFlow;
            capacity[v][u] += pathFlow;
            v = u;
        }

        flow += pathFlow;
    }

    return flow;
}

console.log(maxFlow(1, N));

File: Algorithm/Ford-Fulkerson/Edmonds-Karp/atcoder/A68/A68.py
from collections import deque
import sys

sys.setrecursionlimit(1000000)

# 入力読み込み
input = sys.stdin.read
lines = input().splitlines()
N, M = map(int, lines[0].split())

# グラフ構築
capacity = [[0] * (N + 1) for _ in range(N + 1)]
graph = [[] for _ in range(N + 1)]

for i in range(1, M + 1):
    a, b, c = map(int, lines[i].split())
    capacity[a][b] += c  # 同じ辺が複数あれば合算
    graph[a].append(b)
    graph[b].append(a)  # 逆辺（残余グラフ）も必要

def bfs(s, t, parent):
    visited = [False] * (N + 1)
    queue = deque([s])
    visited[s] = True

    while queue:
        u = queue.popleft()
        for v in graph[u]:
            if not visited[v] and capacity[u][v] > 0:
                parent[v] = u
                visited[v] = True
                if v == t:
                    return True
                queue.append(v)
    return False

def edmonds_karp(s, t):
    flow = 0
    parent = [-1] * (N + 1)

    while bfs(s, t, parent):
        # 最小残余容量を探す
        path_flow = float('inf')
        v = t
        while v != s:
            u = parent[v]
            path_flow = min(path_flow, capacity[u][v])
            v = u

        # 残余容量更新（フロー反映）
        v = t
        while v != s:
            u = parent[v]
            capacity[u][v] -= path_flow
            capacity[v][u] += path_flow
            v = u

        flow += path_flow

    return flow

# 実行：1 から N へ
print(edmonds_karp(1, N))
